drivers/pci/pci.c

```diff
@@ -206,6 +207,30 @@ void LpcSetSerialState(int enable)
 	LpcWriteRegister(0x30, enable ? 0x01 : 0x00);
 }
 
+int serial_putchar(int ch)
+{
+	/* Wait for THRE (bit 5) to be high */
+	while ((IoInputByte(SERIAL_PORT + SERIAL_LSR) & (1 << 5)) == 0);
+	IoOutputByte(SERIAL_PORT + SERIAL_THR, ch);
+	return ch;
```

Comment:
If outputting strings to a serial terminal, you'll likely want to print carriage returns (`\r`) whenever a linefeed character (`\n`) is emitted. Without the carriage return, the terminal may move to the next line and continue at the same place as before. Recommend simply adding a recursive call here.

Reply:
Improved.